app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshakeAsyncTask.kt

```diff
@@ -0,0 +1,189 @@
+            val `object` = objectInputStream.readObject()
+
+            // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+            return@doInBackground if (isKiwixHandshake(`object`) && !isCancelled) {
+              if (BuildConfig.DEBUG) {
+                Log.d(
+                  TAG,
+                  "Client IP address: " + server.inetAddress
+                )
+              }
+              exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream())
+              server.inetAddress
+            } else { // Selected device is not accepting wifi direct connections through the kiwix app
+              null
+            }
+          }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return null
+      }
+    } else if (wifiDirectManager.isGroupFormed && !isCancelled) { //&& !groupInfo.isGroupOwner
+      try {
+        Socket().use { client ->
+          client.reuseAddress = true
+          client.connect(
+            InetSocketAddress(
+              wifiDirectManager.groupOwnerAddress.hostAddress,
+              PEER_HANDSHAKE_PORT
+            ), 15000
+          )
+          val objectOutputStream =
+            ObjectOutputStream(client.getOutputStream())
+          objectOutputStream.writeObject(
+            HANDSHAKE_MESSAGE
+          ) // Send message for the peer device to verify
+          exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream())
+          return@doInBackground wifiDirectManager.groupOwnerAddress
+        }
+      } catch (ex: Exception) {
+        ex.printStackTrace()
+        return null
+      }
+    }
+    return null
+  }
+
+  private fun isKiwixHandshake(`object`: Any): Boolean {
+    return HANDSHAKE_MESSAGE == `object`
+  }
+
+  private fun exchangeFileTransferMetadata(
+    outputStream: OutputStream,
+    inputStream: InputStream
+  ) {
+    if (wifiDirectManager.isFileSender) {
+      try {
+        ObjectOutputStream(outputStream).use { objectOutputStream ->
+          // Send total number of files which will be transferred
+          objectOutputStream.writeObject("" + wifiDirectManager.totalFilesForTransfer)
+          val fileItemArrayList =
+            wifiDirectManager.getFilesForTransfer()
+          for (fileItem in fileItemArrayList) { // Send the names of each of those files, in order
+            objectOutputStream.writeObject(fileItem.fileName)
+            Log.d(
+              TAG,
+              "Sending " + fileItem.fileUri.toString()
+            )
+          }
+        }
+      } catch (e: Exception) {
+        e.printStackTrace()
+      }
+    } else { // Device is not the file sender
+      try {
+        ObjectInputStream(inputStream).use { objectInputStream ->
+          // Read the number of files
+          val totalFilesObject = Integer.parseInt(objectInputStream.readObject().toString())
+          if (totalFilesObject.javaClass == String::class.java) {
```

Comment:
this is undoubtedly an error, how could it be a string after you have parsed it into an Integer?

Reply:
I'm sorry for the variable name for confusing. It must be the number of files from the sender.

Reply:
which probably makes sense for
```
8323-8725/org.kiwix.kiwixmobile D/ReceiverDeviceAsyncTask: Expecting 0 files
8323-8323/org.kiwix.kiwixmobile D/WifiDirectManager: ReceiverDeviceAsyncTask complete
```
The receiver expects 0 files so it completes instantly

Reply:
D'oh! How could i miss that!

Reply:
happens to the best of us, these janky streams are not easy to read